src/client/auth.js

```diff
@@ -1,16 +1,67 @@
 const UUID = require('uuid-1345')
 const yggdrasil = require('yggdrasil')
+const fs = require('mz/fs')
+const promisify = require('es6-promisify')
+const mkdirp = promisify(require('mkdirp'))
 
-module.exports = function (client, options) {
+module.exports = async function (client, options) {
   const yggdrasilClient = yggdrasil({ agent: options.agent, host: options.authServer || 'https://authserver.mojang.com' })
-  const clientToken = options.clientToken || (options.session && options.session.clientToken) || UUID.v4().toString()
+  const clientToken = (options.profilesFolder && (await getLP()).clientToken) || options.clientToken || (options.session && options.session.clientToken) || UUID.v4().toString().replace("-", "")
   const skipValidation = false || options.skipValidation
   options.accessToken = null
-  options.haveCredentials = options.password != null || (clientToken != null && options.session != null)
+  options.haveCredentials = options.password != null || (clientToken != null && options.session != null) || options.profilesFolder != null
+
+  async function getLP () { // get launcher profile
+    let data
+    try {
+      data = JSON.parse(await fs.readFile(options.profilesFolder + '/launcher_profiles.json', 'utf8'))
+    } catch (err) {
+      await mkdirp(options.profilesFolder)
+      await fs.writeFile(options.profilesFolder + '/launcher_profiles.json', '')
+      data = { authenticationDatabase: {} }
+    }
+    return data
+  }
 
   if (options.haveCredentials) {
     // make a request to get the case-correct username before connecting.
-    const cb = function (err, session) {
+    const cb = async function (err, session) {
```

Comment:
I don't like the idea to have an async callback function since this means that the error will be unhandled

Reply:
It's gonna be hard not to have it, or we'll have to use readFileSync and writeFileSync in getLauncherProfiles

Reply:
The problem is not having it but more that the way that it's used here if you have an error in this async function no one is listening to it, I don't think using readFileSync and writeFileSync is the good solution here, maybe just extra the async function elsewhere and use the `.then` on it and handle the error